### PR TITLE
Add SQL CLI project initialisation (flow init)

### PR DESCRIPTION
**Pull request by tatiana**
Closes: #829

WIP:
* [ ] add examples of SQL files, database and connection files (I'm tidying them up )
* [ ] confirm we're not overriding existing files if init is applied to a directory which already had them defined

### CODE REVIEW

File: sql-cli/sql_cli/__main__.py [hunk 1]
```diff
@@ -1,20 +1,23 @@
...
+app = typer.Typer(add_completion=False)
 
 
 @app.command()
 def version() -> None:
     """
     Print the SQL CLI version.
     """
-    rprint("Astro SQL CLI", __version__)
+    rprint(f"Astro SQL CLI {sql_cli.__version__}")
```
Comment (feluelle) [inline sql-cli/sql_cli/__main__.py:20]: I think these changes are unrelated.

File: sql-cli/sql_cli/__main__.py [hunk 2]
```diff
@@ -38,5 +41,53 @@ def generate(
     rprint("The DAG file", dag_file.resolve(), "has been successfully generated. 🎉")
 
 
+@app.command()
+def init(project_dir: Optional[str] = typer.Argument(None)) -> None:
```
Comment (feluelle) [inline sql-cli/sql_cli/__main__.py:45]: Can we use future annotations? And do we need the `typer.Argument` if you don't set it?
```suggestion
def init(project_dir: str | None = None) -> None:
```

File: sql-cli/sql_cli/__main__.py [hunk 3]
```diff
@@ -38,5 +41,53 @@ def generate(
...
+    """
+    Initialise a SQL CLI project structure.
+
+    By default, this includes:
+
+    \b\n
```
Comment (feluelle) [inline sql-cli/sql_cli/__main__.py:51]: I think using `\b` instead of `\b\n` should also work.

File: sql-cli/sql_cli/__main__.py [hunk 4]
```diff
@@ -38,5 +41,53 @@ def generate(
...
+        └── joint_orders_customers.sql
+
+    \b\n
+    Update the file `config/default/configuration.yaml` to declare your databases.
+    \b\n
+    Create SQL workflows within the `workflows` folder.
+    """
+    if project_dir is None:
+        project_dir = os.getcwd()
```
Comment (feluelle) [inline sql-cli/sql_cli/__main__.py:86]:
```suggestion
   project_dir = project_dir or Path.cwd()
```

File: sql-cli/sql_cli/__main__.py [hunk 5]
```diff
@@ -38,5 +41,53 @@ def generate(
...
+
+    Project.initialise(Path(project_dir))
```
Comment (feluelle) [inline sql-cli/sql_cli/__main__.py:88]:
```suggestion
    project = Project(project_dir)
```
Can we not initialise in the `Project` init?

File: sql-cli/sql_cli/__main__.py [hunk 6]
```diff
@@ -38,5 +41,53 @@ def generate(
...
+    rprint(f"Initialized an Astro SQL project at {project_dir}")
```
Comment (feluelle) [inline sql-cli/sql_cli/__main__.py:89]:
```suggestion
    rprint("Initialized an Astro SQL project at", project_dir)
```
I personally use `print(*args)`/`rprint(*args)` feature. You can pass all the values as separate args and print will join it for you and add a space inbetween.

File: sql-cli/sql_cli/__main__.py (hunk 2 above)
Comment (feluelle) [reply sql-cli/sql_cli/__main__.py:45]: Also, you can use `Path` as type. Typer even has [additional features](https://typer.tiangolo.com/tutorial/parameter-types/path/) for it integrated. See [Path Validations](https://typer.tiangolo.com/tutorial/parameter-types/path/#path-validations).

File: sql-cli/sql_cli/project.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import os
+import shutil
+from pathlib import Path
+
+BASE_SOURCE_DIR = Path(Path(os.path.realpath(__file__)).parent.parent, "include/base/")
```
Comment (feluelle) [inline sql-cli/sql_cli/project.py:5]:
```suggestion
BASE_SOURCE_DIR = Path(__file__).parent.parent / "include/base"
```
doesn't this work as well?

File: sql-cli/tests/test___main__.py [hunk 8]
```diff
@@ -20,7 +23,7 @@ def get_stdout(result) -> str:
 def test_about():
     result = runner.invoke(app, ["about"])
     assert result.exit_code == 0
-    assert "Find out more: https://github.com/astronomer/astro-sdk/sql-cli" == get_stdout(result)
+    assert "Find out more" in get_stdout(result)
```
Comment (feluelle) [inline sql-cli/tests/test___main__.py:26]: Why did you change this?

File: sql-cli/tests/test_project.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+    :returns: Sorted list of files and directories within the given directory.
+
+    """
+    return sorted([path.relative_to(dir_name) for path in Path(dir_name).rglob("*")])
+
+
+def test_initialise_project_with_dirname():
+    with tempfile.TemporaryDirectory() as dir_name:
+        assert not os.listdir(dir_name)
```
Comment (feluelle) [inline sql-cli/tests/test_project.py:23]: I personally prefer `pathlib` instead of `os` for path operations. I think all of what you are doing can be rewritten with pathlib :)

File: sql-cli/sql_cli/__main__.py (hunk 2 above)
Comment (tatiana) [reply sql-cli/sql_cli/__main__.py:45]: That's awesome, I love it, I'll use this!

File: sql-cli/sql_cli/__main__.py (hunk 3 above)
Comment (tatiana) [reply sql-cli/sql_cli/__main__.py:51]: It didn't work - that was my first attempt. ATM the code looks disgusting - but the help looks okay:
```
$ poetry run flow init --help
                                                                                                         
 Usage: flow init [OPTIONS] [PROJECT_DIR]                                                                
                                                                                                         
 Initialise a SQL CLI project structure.                                                                 
 By default, this includes:                                                                              
                                                                                                         
                                                                                                         
 ├── config                                                                                              
 │   ├── default                                                                                         
 │   │   └── configuration.yml                                                                           
 │   └── dev                                                                                             
 │       └── configuration.yml                                                                           
 ├── data                                                                                                
 │   ├── movies.db                                                                                       
 │   └── retail.db                                                                                       
 └── workflows                                                                                           
 ├── example_basic_transform                                                                             
 │   └── top_animations.sql                                                                              
 └── example_templating                                                                                  
 ├── filtered_orders.sql                                                                                 
 └── joint_orders_customers.sql                                                                          
                                                                                                         
                                                                                                         
 Update the file `config/default/configuration.yaml` to declare your databases.                          
 Create SQL workflows within the `workflows` folder.                                                     
                                                                                                         
╭─ Arguments ───────────────────────────────────────────────────────────────────────────────────────────╮
│   project_dir      [PROJECT_DIR]  [default: None]                                                     │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────╯
╭─ Options ─────────────────────────────────────────────────────────────────────────────────────────────╮
│ --help          Show this message and exit.                                                           │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────╯
```
I'll follow your advice to see if we can get the directory tree using rich :)

Comment (tatiana) [reply sql-cli/sql_cli/__main__.py:51]: https://rich.readthedocs.io/en/stable/tree.html

File: sql-cli/sql_cli/__main__.py (hunk 5 above)
Comment (tatiana) [reply sql-cli/sql_cli/__main__.py:88]: We can - it didn't matter at the moment, but I can set that, it makes more sense from an OO perspective

File: sql-cli/tests/test___main__.py (hunk 8 above)
Comment (tatiana) [reply sql-cli/tests/test___main__.py:26]: This was a poor rebasing merge - I'll rollback it.

File: sql-cli/sql_cli/__main__.py (hunk 4 above)
Comment (tatiana) [reply sql-cli/sql_cli/__main__.py:86]: From a code coverage perspective, the version before is better ;)

File: sql-cli/sql_cli/__main__.py (hunk 3 above)
Comment (tatiana) [reply sql-cli/sql_cli/__main__.py:75]: @feluelle for now I simplified the docstring - we can improve it in future PRs.

File: sql-cli/pyproject.toml [hunk 10]
```diff
@@ -64,6 +64,6 @@ strict_equality = true
 show_error_codes = true
 show_error_context = true
 warn_redundant_casts = true
-warn_return_any = true
+#warn_return_any = true
```
Comment (feluelle) [inline sql-cli/pyproject.toml:67]:
```suggestion
warn_return_any = true
```
This is unrelated. I had issues with that too, but I would keep the discussion about mypy out of this PR.

File: sql-cli/sql_cli/__main__.py [hunk 11]
```diff
@@ -51,5 +51,59 @@ def validate(environment: str = "default", connection: Optional[str] = None) ->
...
+    Next steps:
+    \b\n
+    * Update the file `config/default/configuration.yaml` to declare database connections.
+    \b\n
+    * Create SQL workflows within the `workflows` folder.
+    """
+    project_dir = project_dir or Path(os.getcwd())
+
+    proj = project.Project(project_dir)
+    proj.initialise(airflow_home, airflow_dags_folder)
```
Comment (feluelle) [inline sql-cli/sql_cli/__main__.py:104]: I like that more. 👍 I realised after talking to you about that, that the project init with creating dirs, and so on is to much for "hiding" it / doing it in the init.

File: sql-cli/sql_cli/__main__.py [hunk 12]
```diff
@@ -51,5 +51,59 @@ def validate(environment: str = "default", connection: Optional[str] = None) ->
...
+    └── workflows: directory where SQL workflows are declared, by default has two examples of workflow
+
+    \b\n
+    Next steps:
+    \b\n
+    * Update the file `config/default/configuration.yaml` to declare database connections.
+    \b\n
+    * Create SQL workflows within the `workflows` folder.
+    """
+    project_dir = project_dir or Path(os.getcwd())
```
Comment (feluelle) [inline sql-cli/sql_cli/__main__.py:101]:
```suggestion
    project_dir = project_dir or Path.cwd()
```

File: sql-cli/sql_cli/project.py [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+        """
+        Delete files which should not be part of the user project directory structure.
+        An example are the .gitkeep files, which were added just so the default folder structure
+        can be added to Git (by default empty directories are not versioned).
+        """
+        gitkeep_files = Path(self.directory).rglob(".gitkeep")
+        for file_ in gitkeep_files:
+            file_.unlink()
+
+    def _update_config(self, airflow_home: Optional[Path], airflow_dags_folder: Optional[Path]) -> None:
```
Comment (feluelle) [inline sql-cli/sql_cli/project.py:33]: Why not use the `self._airflow_dags_folder` and `self._airflow_home` ?

File: sql-cli/sql_cli/project.py [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+        :param airflow_dags_folder: Custom user-defined Airflow DAGs folder
+        """
+        config = Config(environment=DEFAULT_ENVIRONMENT, project_dir=self.directory)
+        if airflow_home is not None:
+            config.write_value_to_yaml("airflow", "home", str(airflow_home))
+        if airflow_dags_folder is not None:
+            config.write_value_to_yaml("airflow", "dags_folder", str(airflow_dags_folder))
+
+    def initialise(
+        self, airflow_home: Optional[Path] = None, airflow_dags_folder: Optional[Path] = None
```
Comment (feluelle) [inline sql-cli/sql_cli/project.py:47]: Same here.

File: sql-cli/tests/test___main__.py [hunk 15]
```diff
@@ -47,3 +50,33 @@ def test_generate(root_directory, dags_directory):
...
+    with tempfile.TemporaryDirectory() as dir_name:
+        result = runner.invoke(app, ["init", dir_name])
+        assert result.exit_code == 0
+        expected_msg = f"Initialized an Astro SQL project at {dir_name}"
+        assert expected_msg in get_stdout(result)
+        assert list_dir(dir_name)
+
+
+def test_init_with_custom_airflow_config():
+    with tempfile.TemporaryDirectory() as dir_name:
```
Comment (feluelle) [inline sql-cli/tests/test___main__.py:65]: Can we not use `runner.isolated_filesystem()` here as well instead of `tempfile.TemporaryDirectory()`?

File: sql-cli/tests/test_project.py [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+BASE_PATHS = [obj for obj in list_dir(str(BASE_SOURCE_DIR)) if not obj.stem == ".gitkeep"]
+
+
+def test_initialise_project_with_dirname():
+    with tempfile.TemporaryDirectory() as dir_name:
+        assert not list_dir(dir_name)
+        Project(dir_name).initialise()
+        new_dir_files_list = list_dir(dir_name)
+        assert new_dir_files_list == BASE_PATHS
+        assert len(new_dir_files_list) == BASE_PATHS_COUNT
```
Comment (feluelle) [inline sql-cli/tests/test_project.py:16]:
```suggestion
        assert len(new_dir_files_list) == len(BASE_PATHS)
```
Why not do this^?

File: sql-cli/sql_cli/__main__.py (hunk 11 above)
Comment (tatiana) [reply sql-cli/sql_cli/__main__.py:104]: Agreed, @feluelle , we can have a method to check if the given directory is a SQL CLI project - I'll be doing this in the next PR, when I align a bit more the validate and init!

File: sql-cli/sql_cli/project.py (hunk 13 above)
Comment (tatiana) [reply sql-cli/sql_cli/project.py:33]: We could, but here I prefer to be explicit about which config properties may be changed.

File: sql-cli/tests/test___main__.py (hunk 15 above)
Comment (tatiana) [reply sql-cli/tests/test___main__.py:65]: The isolated filesystem changes the current directory to be the temporary one being created. I didn't want this behaviour in this test - I wanted the current directory to be pointing to a different folder.

File: sql-cli/tests/test_project.py (hunk 16 above)
Comment (tatiana) [reply sql-cli/tests/test_project.py:16]: Since we are getting the base directory from a configuration, I was worried that by mistake, we change it to an empty directory, and things would stop working. I thought having a counter for how many files/directories we want to copy would help avoid this kind of scenario.

File: sql-cli/pyproject.toml (hunk 10 above)
Comment (tatiana) [reply sql-cli/pyproject.toml:67]: I suggest we address this in a separate PR, there are more urging matters

**Review by feluelle — APPROVED** [review]
